tfhe/src/c_api/high_level_api/utils.rs

```diff
@@ -274,6 +274,53 @@ macro_rules! impl_binary_assign_fn_on_type {
     };
 }
 
+/// The C Standard only define integers fro u8 to u64.
+/// So to support u128 and u256 we had to create our own C friendly
+/// data types.
+///
+/// This trait exists to be able to easily write wrapper function
+/// by allown to generically go from a C API scalar type to a rust one
+pub(in crate::c_api::high_level_api) trait ToRustScalarType {
+    type RustScalarType;
+
+    fn to_rust_scalar_type(self) -> Self::RustScalarType;
+}
+
+/// Implements the trait for when the C API type is the same
+/// as the Rust API type (eg u64)
+macro_rules! impl_to_rust_scalar_type(
+    ($type:ty) => {
+        impl ToRustScalarType for $type {
+            type RustScalarType = $type;
+
+            fn to_rust_scalar_type(self) -> Self::RustScalarType {
+                self
+            }
+        }
+    }
+);
+
+impl_to_rust_scalar_type!(u8);
+impl_to_rust_scalar_type!(u16);
+impl_to_rust_scalar_type!(u32);
+impl_to_rust_scalar_type!(u64);
+
+impl ToRustScalarType for crate::c_api::high_level_api::u128::U128 {
+    type RustScalarType = u128;
+
+    fn to_rust_scalar_type(self) -> Self::RustScalarType {
+        u128::from(self)
+    }
+}
+
+impl ToRustScalarType for crate::c_api::high_level_api::u256::U256 {
+    type RustScalarType = crate::integer::U256;
+
+    fn to_rust_scalar_type(self) -> Self::RustScalarType {
+        crate::integer::U256::from(self)
+    }
+}
```

Comment:
I think I'm missing something here, how does the conversion happen ?

Reply:
which one ? 

In the macro we do
```
let rhs = <$scalar_type as $crate::c_api::high_level_api::utils::ToRustScalarType>::to_rust_scalar_type(rhs);
```

Reply:
I mean there must be a From implementation between the C-friendly type and the rust type ?

Reply:
Yes they are in the C API module

- https://github.com/zama-ai/tfhe-rs/blob/6a56af0b074bfe479221857968ed769a65509a8b/tfhe/src/c_api/high_level_api/u256.rs#L24
- https://github.com/zama-ai/tfhe-rs/blob/main/tfhe/src/c_api/high_level_api/u128.rs

Reply:
thanks !